src/audio/aria/aria_generic.c

```diff
@@ -7,50 +7,93 @@
 
 #ifdef ARIA_GENERIC
 
+/**
+ * \brief Aria gain index mapping table
+ */
+const uint8_t INDEX_TAB[] = {
+		0,    1,    2,    3,
+		4,    5,    6,    7,
+		8,    9,    0,    1,
+		2,    3,    4,    5,
+		6,    7,    8,    9,
+		0,    1,    2,    3
+};
+
 inline void aria_algo_calc_gain(struct comp_dev *dev, size_t gain_idx,
-				int32_t *__restrict data, const size_t src_size)
+				struct audio_stream __sparse_cache *source, int frames)
 {
 	struct aria_data *cd = comp_get_drvdata(dev);
 	int32_t max_data = 0;
-	int64_t gain;
+	uint32_t att = cd->att;
+	int64_t gain = (1ULL << (att + 32)) - 1;
+	int samples = frames * audio_stream_get_channels(source);
+	int32_t *src = audio_stream_get_rptr(source);
+	int i, n;
 
-	for (size_t i = 0; i < src_size; ++i)
-		max_data = MAX(max_data, ABS(data[i]));
+	while (samples) {
+		n = audio_stream_samples_without_wrap_s32(source, src);
+		n = MIN(samples, n);
+		for (i = 0; i < n; i++)
+			max_data = MAX(max_data, ABS(src[i]));
```

Comment:
I'm not sure but is there risk for ABS() to be evaluated twice due to implementation of MAX() macro?

Reply:
from the xtensa test result, it is evaluated twice, thanks!